agents/dqn.py

```diff
@@ -1,137 +1,132 @@
+            if 'optimizer' in config else None
         )
 
     def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
         # Shape: [batch_size]
         current_node_idx = inputs[self._current_node_idx_prefix]
+        if self._node_id < 0:
+            self._node_id = current_node_idx[0].item()
+        batch_size = len(current_node_idx)
         # Shape: [batch_size]
         destination_node_idx = inputs[self._destination_node_idx_prefix]
         # TensorWithMask
         # Shape: [batch_size, max_neighbors_num]
         neighbor_node_ids = inputs[self._neighbors_node_ids_prefix]
 
-        # Shape: [batch_size, max_neighbors_num]
-        next_neighbor_q = self._q_network(
+        # Shape: [batch_size], [batch_size, max_neighbors_num]
+        next_neighbor_ids, neighbors_q = self._q_network(
             current_node_idx=current_node_idx,
             neighbor_node_ids=neighbor_node_ids,
             destination_node_idx=destination_node_idx
         )
 
-        # Shape: [batch_size]
-        best_next_neighbor_idx = neighbor_node_ids.padded_values[torch.argmax(next_neighbor_q, dim=-1)]
-        next_neighbor_idx = _with_random_research(
-            best_next_neighbor_idx,
-            neighbor_node_ids,
-            self._research_prob
-        )
-
         if self._bag_trajectory_memory is not None:
             # TODO[Vladimir Baikalov]: Think about how to generalize
-            bag_ids = inputs.get(self._bag_idx_prefix, None)
+            bag_ids = inputs.get(self._bag_ids_prefix, None)
 
             self._bag_trajectory_memory.add_to_trajectory(
                 bag_ids=bag_ids,
-                node_idxs=current_node_idx,
+                node_idxs=torch.full([batch_size], self._node_id),
                 extra_infos=zip(
-                    current_node_idx,
+                    torch.unsqueeze(current_node_idx, dim=1),
                     neighbor_node_ids,
-                    destination_node_idx,
-                    next_neighbor_idx
+                    neighbors_q,
+                    torch.unsqueeze(destination_node_idx, dim=1),
+                    torch.unsqueeze(next_neighbor_ids, dim=1)
                 )
             )
-        return {
-            'predicted_next_node_idx': next_neighbor_idx,
-            'predicted_next_node_q': next_neighbor_q,
-        }
+        inputs[self._output_prefix] = next_neighbor_ids
+        inputs.update({
+            'predicted_next_node_idx': next_neighbor_ids,
+            'predicted_next_node_q': neighbors_q,
+        })
+        return inputs
 
-    def learn(self):
-        for trajectory in self._bag_trajectory_memory.sample_trajectories_for_node_idx(
-                self._node_idx,
-                self._trajectory_sample_size
-        ):
+    def learn(self) -> Optional[Tensor]:
+        loss = 0
+        learn_trajectories = self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+            self._node_id,
```

Comment:
Let's use keywords here for arguments as an explicit approach. For example, it is unclear what `1` stands for.

Reply:
ok